## Problem Statement

### Business Context

The healthcare industry is rapidly evolving, with professionals facing increasing challenges in managing vast volumes of medical data while delivering accurate and timely diagnoses. The need for quick access to comprehensive, reliable, and up-to-date medical knowledge is critical for improving patient outcomes and ensuring informed decision-making in a fast-paced environment.

Healthcare professionals often encounter information overload, struggling to sift through extensive research and data to create accurate diagnoses and treatment plans. This challenge is amplified by the need for efficiency, particularly in emergencies, where time-sensitive decisions are vital. Furthermore, access to trusted, current medical information from renowned manuals and research papers is essential for maintaining high standards of care.

To address these challenges, healthcare centers can focus on integrating systems that streamline access to medical knowledge, provide tools to support quick decision-making, and enhance efficiency. Leveraging centralized knowledge platforms and ensuring healthcare providers have continuous access to reliable resources can significantly improve patient care and operational effectiveness.

**Common Questions to Answer**

**1. Diagnostic Assistance**: "What are the common symptoms and treatments for pulmonary embolism?"

**2. Drug Information**: "Can you provide the trade names of medications used for treating hypertension?"

**3. Treatment Plans**: "What are the first-line options and alternatives for managing rheumatoid arthritis?"

**4. Specialty Knowledge**: "What are the diagnostic steps for suspected endocrine disorders?"

**5. Critical Care Protocols**: "What is the protocol for managing sepsis in a critical care unit?"

### Objective

As an AI specialist, your task is to develop a RAG-based AI solution using renowned medical manuals to address healthcare challenges. The objective is to **understand** issues like information overload, **apply** AI techniques to streamline decision-making, **analyze** its impact on diagnostics and patient outcomes, **evaluate** its potential to standardize care practices, and **create** a functional prototype demonstrating its feasibility and effectiveness.

### Data Description

The **Merck Manuals** are medical references published by the American pharmaceutical company Merck & Co., that cover a wide range of medical topics, including disorders, tests, diagnoses, and drugs. The manuals have been published since 1899, when Merck & Co. was still a subsidiary of the German company Merck.

The manual is provided as a PDF with over 4,000 pages divided into 23 sections.

## Installing and Importing Necessary Libraries and Dependencies

In [ ]:
# Installation for GPU llama-cpp-python
# uncomment and run the following code in case GPU is being used
!CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python==0.2.28 --force-reinstall --no-cache-dir -q

# Installation for CPU llama-cpp-python
# uncomment and run the following code in case GPU is not being used
# !CMAKE_ARGS="-DLLAMA_CUBLAS=off" FORCE_CMAKE=1 pip install llama-cpp-python==0.2.28 --force-reinstall --no-cache-dir -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.4/9.4 MB 188.3 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 213.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.6/16.6 MB 112.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.6/44.6 kB 254.7 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.19.0 requires numpy<2.2.0,>=1.26.0, but you have numpy 2.4.3 which is incompatible.
numba 0.60.0 requires numpy<2.1,>=1.22, but you have numpy 2.4.3 which is incompatible.


**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

In [ ]:
# For installing the libraries & downloading models from HF Hub
!pip install huggingface_hub==0.35.0 pandas==2.2.2 tiktoken==0.11.0 pymupdf==1.26.4 langchain==0.3.27 langchain-community==0.3.29 chromadb==1.1.0 sentence-transformers==5.1.0 numpy==2.3.3 -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.1/62.1 kB 6.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.0/52.0 kB 5.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.0/44.0 kB 4.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 563.4/563.4 kB 6.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 53.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.1/24.1 MB 41.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 63.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 91.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.8/19.8 MB 31.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 483.4/483.4 kB 40.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.6/16.6 MB 103.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.2/278.2 kB 26.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2

**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

In [ ]:
#Libraries for processing dataframes,text
import json,os
import tiktoken
import pandas as pd


#Libraries for Loading Data, Chunking, Embedding, and Vector Databases
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.document_loaders import PyMuPDFLoader
from langchain_community.embeddings.sentence_transformer import SentenceTransformerEmbeddings
from langchain_community.vectorstores import Chroma

#Libraries for downloading and loading the llm
from huggingface_hub import hf_hub_download
from llama_cpp import Llama

## Question Answering using LLM

#### Downloading and Loading the model

In [ ]:
model_name_or_path = "TheBloke/Mistral-7B-Instruct-v0.2-GGUF"
model_basename = "mistral-7b-instruct-v0.2.Q6_K.gguf"

In [ ]:
model_path = hf_hub_download(
    repo_id=model_name_or_path,
    filename=model_basename
)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


mistral-7b-instruct-v0.2.Q6_K.gguf:   0%|          | 0.00/5.94G [00:00<?, ?B/s]

In [ ]:
llm = Llama(
    model_path=model_path,
    n_ctx=5000,
    n_gpu_layers=38,
    n_batch=512
)

AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 1 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | 


#### Response

In [ ]:
def response(query,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    model_output = llm(
      prompt=query,
      max_tokens=max_tokens,
      temperature=temperature,
      top_p=top_p,
      top_k=top_k
    )

    return model_output['choices'][0]['text']

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
response('What is the protocol for managing sepsis in a critical care unit?')

Llama.generate: prefix-match hit


'\n\nSepsis is a life-threatening condition that can arise from an infection, and it requires prompt recognition and aggressive management in a critical care unit. The following are general steps for managing sepsis in a critical care unit:\n\n1. Early recognition: Recognize the signs and symptoms of sepsis early and initiate treatment as soon as possible. Sepsis can present with various clinical features, including fever or hypothermia, tachycardia or bradycardia, altered mental status, respiratory distress, and lactic acidosis.\n2. ABCs'

In [ ]:
response('What is the protocol for managing sepsis in a critical care unit?',max_tokens=350,temperature= 0.8)

'\n\nSepsis is a life-threatening condition that can arise from an infection, and early recognition and aggressive management are essential for improving outcomes. The following are key components of sepsis management in a critical care unit:\n\n1. Recognition and diagnosis: Septic shock is a subset of sepsis where there is persistent hypotension despite adequate fluid resuscitation. Early recognition and prompt initiation of treatment can improve outcomes. Look for signs of infection, organ dysfunction, and laboratory markers such as elevated white blood cell count, lactate level, and procalcitonin.\n2. Fluid resuscitation: Aggressive fluid resuscitation is the initial step in managing sepsis. The goal is to maintain adequate tissue perfusion, and this may require the administration of large volumes of crystalloids or colloids. Central venous pressure monitoring can be helpful in guiding fluid therapy.\n3. Vasopressor support: If fluid resuscitation fails to restore adequate blood pre

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
response('What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?')

Llama.generate: prefix-match hit


'\n\nAppendicitis is a medical condition characterized by inflammation of the appendix, a small pouch-like structure that extends from the large intestine. The symptoms of appendicitis can vary from person to person, but some common signs include:\n\n1. Abdominal pain: The pain is typically located in the lower right side of the abdomen and may be constant or come and go. It may start as a mild discomfort that worsens over time.\n2. Loss of appetite: People with appendicitis often lose their appetite due to abdominal pain and nausea'

In [ ]:
response('What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?',top_p= 0.2, top_k= 25)

Llama.generate: prefix-match hit


'\n\nAppendicitis is a medical condition characterized by inflammation of the appendix, a small tube-shaped organ located in the lower right side of the abdomen. The symptoms of appendicitis can vary from person to person, but some common signs include:\n\n1. Abdominal pain: The pain is typically located in the lower right quadrant of the abdomen and may start as a mild discomfort that gradually worsens over time. The pain may be constant or intermittent and may be aggravated by movement, deep breathing, or coughing.\n2. Loss of appetite'

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
response('What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?')

Llama.generate: prefix-match hit


"\n\nSudden patchy hair loss, also known as alopecia areata, is a common autoimmune disorder that affects the hair follicles. It can result in round or oval bald patches on the scalp, but it can also occur on other parts of the body such as the beard area, eyebrows, and eyelashes.\n\nThe exact cause of alopecia areata is not known, but it's believed to be related to a problem with the immune system. Some possible triggers for this condition include stress, genetics, viral infections, and certain medications."

In [ ]:
response('What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?',temperature= 0.9, max_tokens= 1024)

Llama.generate: prefix-match hit


"\n\nSudden patchy hair loss, also known as alopecia areata, is a common condition that affects both men and women. It usually presents as round or oval bald patches on the scalp, but it can also occur on other parts of the body such as the beard area, eyebrows, or eyelashes.\n\nThe exact cause of alopecia areata is not known, but it's believed to be an autoimmune condition where the immune system attacks hair follicles, causing them to stop producing new hair. Other potential triggers include stress, genetics, infections, and vitamin deficiencies.\n\nThere are several treatments that can help address sudden patchy hair loss, including:\n\n1. Corticosteroids: These medications can be applied topically or taken orally to reduce inflammation and suppress the immune system's attack on the hair follicles. They have been shown to be effective in promoting hair regrowth in some people with alopecia areata.\n2. Minoxidil: This medication is available over-the-counter and can stimulate hair gr

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
response('What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?')

Llama.generate: prefix-match hit


'\n\nA person who has sustained a physical injury to the brain tissue may require various treatments depending on the severity and location of the injury. Here are some common treatments that may be recommended:\n\n1. Emergency care: In case of a traumatic brain injury (TBI), it is essential to seek emergency medical attention as soon as possible. The primary goal of emergency care is to prevent further damage to the brain, stabilize vital signs, and manage any life-threatening conditions.\n2. Medications: Depending on the symptoms, healthcare professionals may prescribe medications to manage various conditions associated with a'

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
response('What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?')

Llama.generate: prefix-match hit


"\n\nFirst and foremost, if you suspect that someone has fractured their leg while hiking, it's essential to ensure their safety and prevent further injury. Here are some necessary precautions:\n\n1. Keep the person calm and still: Encourage them to remain as still as possible to minimize pain and prevent worsening the injury.\n2. Assess the situation: Check for any signs of shock, such as pale skin, rapid heartbeat, or shallow breathing. If you notice these symptoms, seek medical help immediately.\n3. Immobilize the leg: Use a splint, sl"

**Observations:**

*   The pre trained LLM model responded to the questions as per training data.
*   Tried different parameter such as temparature and max token and saw the different responses from LLM. when we increase the temparature value, the response is more creative then deterministic.
*  With max token size increse, the response text increase too.


## Question Answering using LLM with Prompt Engineering

In [ ]:
def generate_llama_response(instruction, review):

    # System message explicitly instructing not to include the review text
    system_message = """
        [INST]<<SYS>>
        {}
        <</SYS>>[/INST]
    """.format(instruction)

    # Combine user_prompt and system_message to create the prompt
    prompt = f"{review}\n{system_message}"

    # Generate a response from the LLaMA model
    response = llm(
        prompt=prompt,
        max_tokens=1024,
        temperature=0,
        top_p=0.95,
        repeat_penalty=1.2,
        top_k=50,
        stop=['INST'],
        echo=False,
        seed=42,
    )

    # Extract the sentiment from the response
    response_text = response["choices"][0]["text"]
    return response_text

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
generate_llama_response('''Explain the protocol for managing sepsis in a critical care unit. Include early recognition, initial resuscitation steps, antibiotic therapy, hemodynamic support, and ongoing monitoring. Base your answer on current Surviving Sepsis Campaign guidelines.''','What is the protocol for managing sepsis in a critical care unit?')

Llama.generate: prefix-match hit


" The Surviving Sepsis Campaign (SSC) is an international guideline initiative aimed at improving care and outcomes for sepsis patients. Here's a summary of the recommended protocol for managing sepsis in a critical care unit based on current SSC guidelines:\n\n1. Early recognition:\n   - Suspect sepsis if there's suspicion or confirmation of infection, plus signs of organ dysfunction (using Sequential [Sepsis-related] Organ Failure Assessment, SOFA score ≥2 points).\n   - Consider septic shock when there is sepsis with persisting hypotension requiring vasopressors to maintain mean arterial pressure ≥65 mmHg and serum lactate level >2 mmol/L despite adequate volume resuscitation.\n\n2. Initial resuscitation steps:\n   - Give high-flow oxygen via a nonrebreather mask or endotracheal tube to maintain SpO2 ≥96%.\n   - Initiate intravenous fluid resuscitation with crystalloids (30 mL/kg in the first hour, then 3 mL/kg per hour for next 8 hours).\n   - Administer broad-spectrum antibiotics 

In [ ]:
generate_llama_response('''Provide a step-by-step protocol for managing sepsis in an ICU setting. Organize the response into:
1. Initial assessment and diagnosis
2. First-hour bundle (including fluid resuscitation and antibiotics)
3. Hemodynamic management (vasopressors, MAP targets)
4. Source control
5. Monitoring and supportive care

Use evidence-based practices and include key clinical thresholds and targets.''','What is the protocol for managing sepsis in a critical care unit?')

Llama.generate: prefix-match hit


'1. Initial Assessment and Diagnosis:\n       a. Recognize signs of infection (fever, chills, tachycardia, increased respiratory rate) and inflammation (increased WBC count, lactate level).\n       b. Use the Sequential [Sepsis-related] Organ Failure Assessment (SOFA) score to evaluate organ dysfunction. A SOFA score ≥2 indicates sepsis, while a score ≥4 suggests severe sepsis or septic shock.\n       c. Obtain blood cultures before administering antibiotics if possible.\n\n    2. First-hour bundle:\n       a. Fluid resuscitation: Administer intravenous crystalloids to achieve a target mean arterial pressure (MAP) of ≥65 mmHg and a central venous pressure (CVP) between 8-12 cm H2O or a pulmonary capillary wedge pressure (PCWP) between 12-14 mmHg.\n       b. Antibiotics: Administer broad-spectrum antibiotics based on culture results and suspected infection site within the first hour of recognition and resuscitation.\n\n    3. Hemodynamic management:\n       a. Vasopressors: If MAP remai

In [ ]:
generate_llama_response('''Teach a medical student how sepsis is managed in the ICU. Use simple language but include all essential steps such as recognition, resuscitation, antibiotics, and organ support. Add clinical reasoning and common pitfalls.''','What is the protocol for managing sepsis in a critical care unit?')

Llama.generate: prefix-match hit


" Sure thing! Sepsis is a life-threatening condition that can occur when the body's response to an infection becomes uncontrolled and causes damage to various organs. Here are the essential steps for managing sepsis in an ICU setting:\n\n1. Recognition: The first step is recognizing sepsis early, as timely intervention can improve outcomes. Look out for signs of infection such as fever or hypothermia, increased heart rate and respiratory rate, altered mental status, and lactic acidosis. Use the Sequential Organ Failure Assessment (SOFA) score to assess organ dysfunction.\n2. Resuscitation: The next step is to resuscitate the patient to restore perfusion and oxygenation. This includes administering intravenous fluids to maintain adequate blood pressure, providing supplemental oxygen if necessary, and initiating cardiopulmonary support as needed.\n3. Antibiotics: Once sepsis has been diagnosed or suspected, start broad-spectrum antibiotics as soon as possible. The choice of antibiotics w

In [ ]:
generate_llama_response('''Act as an ICU specialist. Describe the evidence-based management of sepsis and septic shock in critically ill patients. Include:
- Pathophysiology overview
- Sepsis-3 diagnostic criteria
- Early goal-directed therapy updates
- Fluid choice and volume recommendations
- Vasopressor selection and titration
- भूमिका of lactate monitoring
- Antibiotic timing and de-escalation
- Organ support (ventilation, renal replacement therapy)

Reference Surviving Sepsis Campaign recommendations where applicable.


''','What is the protocol for managing sepsis in a critical care unit?')

Llama.generate: prefix-match hit


'1. Initial assessment and diagnosis:\n       a. Recognize signs of sepsis: fever or hypothermia, tachycardia or bradycardia, respiratory distress, altered mental status, lactic acidosis.\n       b. Use the Sequential [Sepsis-related] Organ Failure Assessment (SOFA) score to evaluate organ dysfunction. A SOFA score ≥2 indicates sepsis and a score ≥4 suggests severe sepsis or septic shock.\n       c. Obtain blood cultures before administering antibiotics, if possible.\n\n    2. First-hour bundle:\n       a. Fluid resuscitation: give 30 mL/kg of crystalloid within the first hour for patients with sepsis and septic shock. Use colloids or albumin in addition to crystalloids if needed for fluid refractory hypotension.\n       b. Antibiotics: administer broad-spectrum antibiotics based on suspected infection site, microbiology data, and local guidelines within the first hour of recognition and resuscitation.\n\n    3. Hemodynamic management:\n       a. Maintain mean arterial pressure (MAP) ≥

In [ ]:
generate_llama_response('''Create a concise ICU checklist for sepsis management based on current clinical guidelines. Include time-sensitive interventions (e.g., within 1 hour), drug choices, and monitoring parameters.''','What is the protocol for managing sepsis in a critical care unit?')

Llama.generate: prefix-match hit


'1. **Recognition**: Suspect sepsis in any patient with suspected or confirmed infection and organ dysfunction using the Sequential [Sepsis-related Organ Failure Assessment (SOFA)] score ≥2 points increase.\n\n    **Within 1 hour:**\n    1. High-flow oxygen via nasal cannulas, non-rebreather masks, or endotracheal tubes as needed to maintain SpO2 >96%.\n    2. Initiate intravenous (IV) access with large bore catheters and administer 30 mL/kg crystalloid solution if hypotensive or lactate level >4 mmol/L. Repeat this bolus up to three times if needed, then reassess hemodynamic status.\n    3. Obtain blood cultures before starting antibiotics (if not already done).\n    4. Start broad-spectrum IV antibiotics based on suspected infection site and microbiology data.\n    5. Administer IV fluid resuscitation as needed to maintain MAP >65 mmHg or mean arterial pressure (MAP) ≥20% increase from baseline, whichever is higher.\n    6. Monitor vital signs every hour: temperature, heart rate, res

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
generate_llama_response('''Explain the common symptoms of appendicitis and whether it can be treated with medication alone. If surgery is required, describe the standard surgical procedure used to treat it, including indications and basic steps.''','What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?')

Llama.generate: prefix-match hit


' Appendicitis is a medical condition characterized by inflammation of the appendix, a small tube-shaped organ located in the lower right side of the abdomen. The following are common symptoms of appendicitis:\n     1. Abdominal pain: This is usually the first and most prominent symptom. It starts as a dull ache around the navel area but eventually localizes to the lower right quadrant of the abdomen. The pain may be constant or intermittent, worsening with movement or coughing.\n     2. Loss of appetite: As the inflammation progresses, you might lose your appetite and feel nauseous or vomit.\n     3. Fever: A low-grade fever (100°F to 104°F) is common in appendicitis cases.\n     4. Rebound tenderness: When the doctor presses on the affected area, it may cause discomfort and pain that disappears when pressure is released – this is known as rebound tenderness.\n     5. Guarding: The muscles of the abdominal wall might involuntarily contract to protect the inflamed appendix from further

In [ ]:
generate_llama_response('''Provide a structured explanation of appendicitis including:
1. Common signs and symptoms
2. When antibiotic therapy may or may not be sufficient
3. Indications for surgery
4. Description of the standard surgical procedure (appendectomy)

Base your answer on current clinical guidelines and keep it concise but accurate.''','What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?')

Llama.generate: prefix-match hit


"1. Appendicitis is an inflammatory condition of the appendix, a small tube-shaped organ located in the lower right abdomen. Common signs and symptoms include:\n   - Sudden pain around the navel or near the right lower abdomen that may worsen over hours\n   - Loss of appetite\n   - Nausea and vomiting\n   - Fever, often low-grade at first but rising as the condition progresses\n   - Abdominal swelling and tenderness, especially in the right lower quadrant\n   - Diarrhea or constipation\n   - Inability to pass gas or have a bowel movement\n   - Restlessness and feeling ill in general.\n\n2. Antibiotic therapy may be considered as an alternative to surgery for certain cases of uncomplicated appendicitis, such as those with mild symptoms, no signs of perforation (rupture) on imaging studies, or a strong desire not to undergo surgery. However, this approach is associated with higher rates of recurrence and complications compared to surgical intervention.\n\n3. Indications for surgery inclu

In [ ]:
generate_llama_response('''Explain in simple terms what symptoms people usually have with appendicitis. Can it be treated with medicines alone? If not, explain what surgery is done and what it involves, in a way a patient can understand.''','What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?')

Llama.generate: prefix-match hit


" Appendicitis is an inflammation of the appendix, which is a small tube-shaped organ attached to your large intestine. When you have appendicitis, common symptoms include:\n\n1. Pain in the lower right side of your abdomen (stomach area) that may start as mild and then become sharp and severe over time.\n2. Loss of appetite and feeling sick to your stomach.\n3. Fever and chills.\n4. Feeling bloated or swollen in the abdominal area.\n5. Constipation, diarrhea, or both.\n6. Pain when you move, cough, or sneeze.\n7. Inability to pass gas or have a bowel movement.\n\nAppendicitis cannot be cured with medicines alone because it requires surgical removal of the appendix (an appendectomy) due to the risk of rupture and infection spreading throughout your body if left untreated. During an appendectomy, the surgeon makes a small incision in your abdomen to remove the inflamed appendix. The procedure can be done using traditional open surgery or laparoscopically (using smaller incisions and a c

In [ ]:
generate_llama_response('''Act as a general surgeon. Explain the presentation and management of appendicitis. Include:
- Typical and atypical symptoms
- Role of imaging in diagnosis
- Evidence for non-operative (antibiotic) management vs surgical treatment
- Indications for appendectomy
- Differences between open and laparoscopic appendectomy
''','What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?')

Llama.generate: prefix-match hit


" Appendicitis is a common acute inflammatory condition of the vermiform appendix, which is a small, tubular structure that extends from the cecum in the right lower abdomen. The typical presentation of appendicitis includes:\n\n1. Periumbilical or right lower quadrant (RLQ) pain, which may begin as vague discomfort and progress to constant sharp pain over several hours.\n2. Anorexia, nausea, and vomiting are common symptoms.\n3. Fever is often present with an elevated white blood cell count.\n4. Rebound tenderness or guarding on physical examination in the RLQ area.\n5. Loss of appetite (anorexia) and constipation or diarrhea may also occur, but these symptoms are less specific for appendicitis.\n\nAtypical presentations can include:\n- Left lower quadrant pain due to a retrocecal or pelvic appendix.\n- Perforated appendicitis with systemic signs of sepsis and peritonitis.\n- Appendicitis in pregnant women, where the symptoms may be masked by pregnancy-related changes.\n\nImaging play

In [ ]:
generate_llama_response('''Create a concise checklist for appendicitis covering symptoms, diagnosis, when antibiotics are appropriate, and the recommended surgical procedure.''','What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?')

Llama.generate: prefix-match hit


"1. Symptoms: Appendicitis often starts with a pain in the middle of your abdomen that may come and go. The pain eventually becomes constant and may be accompanied by loss of appetite, nausea, vomiting, diarrhea or constipation, low-grade fever, and a feeling of being sick. Pain typically localizes to the lower right side of the abdomen as the condition progresses.\n    2. Diagnosis: A physical exam is usually sufficient for diagnosis, but imaging tests like ultrasound or CT scan may be used if there's uncertainty. These tests can help confirm appendicitis and assess its severity.\n    3. Antibiotics: In some cases where the condition is caught very early on, antibiotics alone might be considered to treat appendicitis. However, this approach carries risks such as rupture of the appendix or failure to effectively eliminate infection. It's generally recommended for those who are not good candidates for surgery due to age, health conditions, or personal preference.\n    4. Surgical Proced

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
generate_llama_response('''Explain the possible causes of sudden patchy hair loss (localized bald spots on the scalp) and outline effective treatment options. Include both medical and non-medical approaches, and clarify whether the condition is reversible.''','What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?')

Llama.generate: prefix-match hit


" Sudden patchy hair loss, also known as alopecia areata or spot baldness, can be a distressing condition that affects people of all ages and genders. The exact cause of this condition is not fully understood, but it's believed to involve an abnormal response of the immune system against the hair follicles. Here are some possible causes:\n\n    1. Genetics: There seems to be a genetic predisposition for developing alopecia areata. If other family members have had this condition, you may be more likely to develop it.\n    2. Immune System Dysfunction: The immune system mistakenly attacks the hair follicles, leading to inflammation and hair loss. Stress can trigger an autoimmune response in some people, causing alopecia areata.\n    3. Vitamin Deficiencies or Nutritional Imbalances: Lack of certain vitamins like biotin, iron, zinc, or protein may contribute to patchy hair loss. However, it's essential to note that these deficiencies usually cause more widespread hair thinning rather than

In [ ]:
generate_llama_response('''Provide a structured explanation of sudden patchy hair loss (e.g., alopecia areata) including:
1. Common causes (autoimmune, nutritional, stress-related, etc.)
2. Typical signs and symptoms
3. Evidence-based treatment options (topical, injectable, systemic therapies)
4. Prognosis and likelihood of hair regrowth

Base your answer on current dermatology guidelines.''','What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?')

Llama.generate: prefix-match hit


"1. Common Causes:\n       a. Alopecia Areata (AA): An autoimmune disorder that causes patchy hair loss, often starting with one or more circular bald spots on the scalp. It can also affect beards and eyebrows. The exact cause is unknown but it's believed to involve an abnormal response of the immune system against melanocytes in the hair follicles.\n       b. Nutritional Deficiencies: Lack of essential nutrients like iron, zinc, biotin, or protein can lead to sudden patchy hair loss. This is usually seen as diffuse thinning rather than distinct patches but it's important to rule out this possibility in cases of alopecia areata as well.\n       c. Stress: Sudden onset of stressful events may trigger an episode of alopecia areata or worsen existing conditions due to the impact of stress hormones on hair growth cycle.\n       d. Other Causes: Certain medications, infections (such as ringworm), trauma, and certain medical conditions like thyroid disease can also cause patchy hair loss.\n\

In [ ]:
generate_llama_response('''Explain in simple terms why someone might suddenly develop small bald patches on their scalp. What are the common causes, and what treatments are available? Can the hair grow back?''','What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?')

Llama.generate: prefix-match hit


" Someone may experience sudden patchy hair loss due to various reasons. The most common causes include:\n\n1. Alopecia Areata: This is an autoimmune condition where your immune system attacks hair follicles, leading to bald patches on the scalp or other areas of the body. It often starts with a single coin-sized spot but can spread and affect larger areas over time.\n2. Trauma or stress: Sudden emotional or physical shock can cause temporary hair loss in small patches. This is known as telogen effluvium, which occurs when hairs prematurely enter the shedding phase due to stressors on the body.\n3. Nutritional deficiencies: Lack of essential nutrients like iron, zinc, or vitamin B12 can lead to hair loss in patches. This is because these nutrients are crucial for maintaining healthy hair growth.\n4. Medications and treatments: Certain medications such as chemotherapy drugs, steroids, or antidepressants may cause patchy hair loss as a side effect.\n5. Fungal infections: Scalp fungal inf

In [ ]:
generate_llama_response('''Act as a dermatologist. Explain the etiology, clinical presentation, and management of patchy hair loss on the scalp. Include:
- Autoimmune mechanisms (e.g., alopecia areata)
- Differential diagnoses (e.g., traction alopecia, nutritional deficiencies)
- Diagnostic approach
- First-line and advanced treatments (corticosteroids, immunotherapy, JAK inhibitors)
- Factors influencing prognosis
 ''','What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?')

Llama.generate: prefix-match hit


" As a dermatologist, I'd be happy to help answer your question about patchy hair loss on the scalp. Patchy hair loss refers to sudden and localized bald spots that can occur for various reasons. Let me first discuss some common causes of this condition:\n\n1. Alopecia Areata (AA): This is an autoimmune disorder where the body's immune system attacks hair follicles, leading to hair loss in circular patches on the scalp or other areas of the body. AA can progress to total baldness (alopecia totalis) or complete loss of all body hair (alopecia universalis).\n\n2. Differential Diagnoses: Several conditions may present with patchy hair loss, including traction alopecia (hair loss due to pulling on the hair), nutritional deficiencies (such as iron-deficiency anemia or zinc deficiency), thyroid disorders, and certain medications.\n\n3. Clinical Presentation: In AA, patchy hair loss typically appears suddenly and progresses over days to weeks. The affected areas may be smooth with no signs of

In [ ]:
generate_llama_response('''Create a concise checklist for patchy hair loss covering causes, symptoms, treatment options, and when to see a doctor.''','What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?')

Llama.generate: prefix-match hit


'1. Causes:\n       * Alopecia Areata (autoimmune disorder)\n       * Trauma or stress\n       * Nutritional deficiencies (iron, protein)\n       * Hormonal imbalances\n       * Certain medications (chemotherapy, thyroid medication)\n       * Infections (ringworm, folliculitis)\n       * Other medical conditions (lupus, syphilis)\n    2. Symptoms:\n       * Round or oval bald patches on the scalp\n       * Sudden hair loss in other areas of the body\n       * Hair thinning around the edges of the patch\n       * Itching, burning sensation, or pain near the affected area\n      3. Treatment options:\n         * Topical treatments (minoxidil, corticosteroids)\n         * Oral medications (corticosteroids, immunosuppressants)\n         * Light therapy (laser treatment)\n         * Alternative therapies (acupuncture, herbal remedies)\n      4. When to see a doctor:\n         * If the patchy hair loss is sudden and unexplained\n         * If you have other symptoms such as fever, weight los

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
generate_llama_response('''Explain the recommended treatments for a person with a traumatic brain injury (TBI), including both immediate emergency care and long-term rehabilitation. Cover mild, moderate, and severe cases.''','What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?')

Llama.generate: prefix-match hit


" Traumatic Brain Injuries (TBIs) can range from mild to severe, each requiring different levels of intervention. Here's an overview of the recommended treatments for TBI patients:\n\n1. Immediate Emergency Care:\n   a. Assessing and Stabilizing Airway, Breathing, and Circulation (ABCs): This is the first priority in managing any trauma patient, including those with TBIs.\n   b. Controlling Bleeding: Applying direct pressure to stop bleeding from wounds or using surgical intervention if necessary.\n   c. Managing Intracranial Pressure (ICP): ICP monitoring and management may be required for severe cases to prevent further damage due to swelling or hemorrhage.\n   d. Administering Oxygen: Supplying adequate oxygen through supplemental sources, such as masks or ventilators, if the patient is not breathing properly.\n   e. Providing Fluid and Electrolyte Balance: Ensuring proper hydration and electrolyte balance to maintain body function.\n\n2. Medications:\n   a. Pain Relievers: Over-the

In [ ]:
generate_llama_response('''Provide a structured overview of traumatic brain injury (TBI) management including:
1. Initial emergency assessment and stabilization (ABC approach)
2. Acute medical/surgical treatments (e.g., ICP management, neurosurgery)
3. Pharmacological management
4. Rehabilitation strategies (physical, cognitive, behavioral therapy)
5. Prognosis and recovery factors

Base your answer on current clinical guidelines.''','What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?')

Llama.generate: prefix-match hit


"1. Initial Emergency Assessment and Stabilization:\n   The first priority in managing a traumatic brain injury (TBI) is to ensure the person's airway, breathing, and circulation are stable using the ABC approach. This includes assessing and maintaining an open airway, providing oxygen support if necessary, and controlling any bleeding or hemorrhage.\n\n2. Acute Medical/Surgical Treatments:\n   a) Intracranial Pressure (ICP) Management: Monitor ICP to prevent further damage due to increased pressure within the skull. This may involve using intraventricular catheters, external ventricular drains, or other devices to manage cerebrospinal fluid and maintain optimal brain tissue volume.\n   b) Neurosurgery: Depending on the severity of the injury, surgical intervention might be required to remove hematomas (collections of blood), repair skull fractures, or relieve pressure within the skull.\n\n3. Pharmacological Management:\n   a) Steroids: Dexamethasone may be used in some cases to reduce

In [ ]:
generate_llama_response('''Explain in simple terms how doctors treat a brain injury caused by trauma. Include what happens in the hospital right after injury and what recovery or rehabilitation may involve.''','What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?')

Llama.generate: prefix-match hit


' After sustaining a traumatic brain injury (TBI), the first priority for medical professionals is to ensure the person\'s airway is clear, they are breathing properly, and their heart rate and blood pressure are stable. This initial assessment and care is often referred to as "primary survey" or "ABCDE approach": Airway, Breathing, Circulation, Disability (neurological status), and Exposure.\n\n     If the person\'s condition is critical, they may be admitted to an intensive care unit (ICU) for close monitoring and supportive treatments such as:\n     1. Oxygen therapy: To help the brain get enough oxygen through a breathing tube or mask.\n     2. Medications: To manage symptoms like increased intracranial pressure, seizures, or decreased consciousness.\n     3. Fluids and nutrients: Intravenous fluids to maintain hydration and prevent dehydration; enteral nutrition if the person cannot eat normally.\n     4. Surgery: Depending on the injury\'s location and severity, surgery may be re

In [ ]:
generate_llama_response('''Act as a neurologist or critical care specialist. Describe the pathophysiology and management of traumatic brain injury (TBI). Include:
- Classification (mild, moderate, severe; Glasgow Coma Scale)
- Neuroimaging and monitoring (CT, ICP monitoring)
- Indications for surgical intervention (e.g., hematoma evacuation, decompressive craniectomy)
- Medical management (sedation, osmotic therapy, seizure prophylaxis)
- Multidisciplinary rehabilitation and long-term care
''','What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?')

Llama.generate: prefix-match hit


" Traumatic Brain Injury (TBI) is caused by external physical forces that disrupt the normal function of the brain. The severity of TBI can range from mild, with transient impairment of consciousness or cognitive functions, to severe, resulting in prolonged coma and permanent disability or even death.\n\n     Classification:\n     1. Mild TBI (also known as concussion): GCS score 15 at presentation, no skull fracture, loss of consciousness less than 30 minutes, and post-traumatic amnesia lasting less than 24 hours.\n     2. Moderate TBI: GCS score between 9 and 12 (awake but confused), or a GCS score of 13-15 with other signs of brain injury such as skull fracture, hemorrhage on CT scan, or post-traumatic amnesia lasting more than 24 hours.\n     3. Severe TBI: GCS score less than 9 (unresponsive), or a GCS score between 9 and 15 with signs of brain herniation, intracranial hematoma, or other life-threatening conditions.\n\n     Neuroimaging and Monitoring:\n     1. Computed Tomography

In [ ]:
generate_llama_response('''Create a concise checklist for traumatic brain injury (TBI) treatment covering emergency care, acute interventions, medications, and rehabilitation steps.''','What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?')

Llama.generate: prefix-match hit


'1. Emergency Care:\n       - Ensure airway is clear and the person is breathing.\n       - Check for signs of respiratory distress or shock.\n       - Immobilize the head and neck to prevent further injury.\n       - Transport the individual to a hospital as soon as possible.\n\n    2. Acute Interventions:\n       - Monitor vital signs, level of consciousness, and neurological status.\n       - Administer oxygen if needed.\n       - Control intracranial pressure (ICP) through various methods such as decompressive craniectomy or hyperventilation.\n       - Manage increased intracranial pressure caused by cerebral edema using osmotic agents, diuretics, and corticosteroids.\n       - Treat seizures with anticonvulsant medications if they occur.\n       - Monitor for signs of infection or bleeding in the brain.\n\n    3. Medications:\n       - Prescribe analgesics to manage pain.\n       - Administer anti-inflammatory drugs to reduce swelling and inflammation.\n       - Provide anticonvul

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
generate_llama_response('''Explain the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip. Include immediate first aid, safe transport, medical treatment, and recovery considerations.''','What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?')


Llama.generate: prefix-match hit


"1. Immediate First Aid:\n    - Ensure the safety of both the injured person and those around them by securing the area if necessary.\n    - Encourage the person to remain calm and still to minimize pain and further injury.\n    - Check for signs of shock, such as pale skin, rapid heartbeat, or shallow breathing. If present, lay the person down with their legs elevated above heart level.\n    - Apply a sterile dressing over any visible wounds to prevent infection. Do not attempt to realign the bone or apply excessive pressure on the fracture site.\n    2. Safe Transport:\n    - Avoid moving the injured person unnecessarily, as this could cause additional pain and complications. If possible, call for emergency medical services (EMS) or arrange for transportation to a hospital.\n    - If transport is required before EMS arrival, use a proper immobilization device such as a splint or a commercially available leg brace. Make sure the person is securely fastened in place during transport an

In [ ]:
generate_llama_response('''Provide a structured guide for managing a suspected leg fracture during a hiking trip
1. Immediate first aid steps in a remote setting
2. Precautions to prevent further injury
3. Immobilization techniques and pain management
4. When and how to evacuate the person safely
5. Hospital treatment options
6. Recovery and rehabilitation considerations

Ensure the advice is practical and suitable for non-medical individuals in outdoor settings.''','What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?')

Llama.generate: prefix-match hit


"1. Immediate first aid steps in a remote setting:\n   a. Assess the situation and ensure your own safety before approaching the injured person.\n   b. Encourage the person to remain calm and still to minimize pain and further injury.\n   c. Check for signs of open wounds, swelling, deformity or numbness in the leg.\n   d. Do not attempt to realign the bone as this could cause more harm.\n   e. Apply a sterile dressing if there is an open wound, but do not remove it unless contaminated.\n   f. Splint the injured limb using available materials such as sticks, branches or hiking poles, padded with clothing to prevent pressure points and maintain circulation. Securely fasten the splint in place using tape, straps or other means.\n   2. Precautions to prevent further injury:\n   a. Keep the person warm by covering them with blankets or extra layers of clothing.\n   b. Provide fluids to help keep them hydrated and offer non-alcoholic beverages if they are able to drink.\n   c. Monitor for s

In [ ]:
generate_llama_response('''Act as an emergency medicine or wilderness medicine expert. Describe the management of a leg fracture sustained during a hiking trip. Include:
- Field assessment (neurovascular status, open vs closed fracture)
- Splinting techniques using improvised materials
- Indications for urgent evacuation vs delayed transport
- Risks (shock, compartment syndrome, infection)
- Definitive treatment (casting, surgical fixation)
- Rehabilitation and return-to-activity guidance
        ''','What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?')

Llama.generate: prefix-match hit


" I. Field Assessment:\n    A. Neurovascular Status: Check for any signs of neurovascular compromise such as loss of sensation, weakness or paralysis, pulselessness, or cold extremity. This is crucial to assess the severity and potential complications of the fracture.\n    B. Open vs Closed Fractures: Determine if the fracture is open (compound) or closed by examining the skin around the injury for any signs of penetration or laceration. An open fracture increases the risk of infection, so prompt medical attention and wound care are essential.\n\n     II. Splinting Techniques:\n    A. Use improvised materials such as sticks, branches, or clothing to create a splint that stabilizes the leg below the injury site while maintaining alignment with the uninjured limb. This will help prevent further damage and provide some pain relief.\n      1. Long-leg splint for femur fractures: Extend the injured leg straight out in front of you, then create a long splint that extends from just above the 

In [ ]:
generate_llama_response('''Explain in simple terms what to do if someone breaks their leg while hiking. Include step-by-step first aid, how to keep them safe until help arrives, and what recovery will look like afterward.''',
                        'What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?')

Llama.generate: prefix-match hit


"1. Stay Calm: The first thing you should do is stay calm and assess the situation. Encourage the injured person to remain still to prevent further injury or worsening of their condition.\n\n2. Call for Help: If possible, call emergency services for medical assistance. Provide them with your location and describe the nature of the injury.\n\n3. First Aid: While waiting for help, provide basic first aid by following these steps:\n   a) Immobilize the leg: Use a splint or any available materials to immobilize the injured leg above and below the fracture site. This will prevent movement and reduce pain.\n   b) Control bleeding: Apply direct pressure on the wound with a clean cloth if there is heavy bleeding. Keep applying pressure until it slows down or stops completely.\n   c) Comfort and warmth: Cover the injury with a sterile dressing, blanket, or insulating material to keep them warm and comfortable.\n\n4. Safety Considerations: Ensure that the injured person is in a safe location awa

In [ ]:
generate_llama_response('''Create a concise checklist for handling a leg fracture in a remote hiking environment, including first aid, precautions, evacuation, medical care, and recovery steps.''',
                       'What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?')

Llama.generate: prefix-match hit


"1. Assess the situation: Check if the person is in immediate danger or requires urgent attention due to other injuries or conditions. If it's safe for you to approach them, proceed with first aid measures.\n\n2. First Aid:\n   a. Immobilize the leg using a splint or improvised materials like sticks and clothing. Make sure the fracture is not displaced (out of alignment) during immobilization.\n   b. Apply a sterile dressing to any open wounds around the fracture site, if present.\n   c. Control bleeding by applying direct pressure on the wound with a clean cloth or bandage.\n   d. Provide pain relief using over-the-counter medications like acetaminophen or ibuprofen (if available).\n\n3. Precautions:\n   a. Avoid moving the person excessively to prevent further injury or worsening of their condition.\n   b. Keep the fractured leg elevated and supported above heart level whenever possible to reduce swelling and pain.\n   c. Monitor for signs of shock, such as rapid pulse, shallow breat

**Observations:**

*  The same question produced different responses depending on how the prompt was engineered.
*  When the prompt was clear and direct, the response was concise and did not include step-by-step guidance.
*  Prompts that included structured steps guided the LLM effectively, and the model followed those steps in its responses.
*  When asked to provide a checklist for a healthcare procedure or surgery, the LLM generated accurate and well-structured answers.

## Data Preparation for RAG

### Loading the Data

In [3]:
# uncomment and run the following lines for Google Colab
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
manual_pdf_path = '/content/drive/MyDrive/MedicalAssistant/medical_diagnosis_manual.pdf'

In [5]:
pdf_loader = PyMuPDFLoader(manual_pdf_path)

In [6]:
manual = pdf_loader.load()

### Data Overview

#### Checking the first 5 pages

In [7]:
for i in range(5):
    print(f"Page Number : {i+1}",end="\n")
    print(manual[i].page_content,end="\n")

Page Number : 1
nilesh.ghodke.usa@gmail.com
H5U9LRDI2E
This file is meant for personal use by nilesh.ghodke.usa@gmail.com only.
Sharing or publishing the contents in part or full is liable for legal action.
Page Number : 2
nilesh.ghodke.usa@gmail.com
H5U9LRDI2E
This file is meant for personal use by nilesh.ghodke.usa@gmail.com only.
Sharing or publishing the contents in part or full is liable for legal action.
Page Number : 3
Table of Contents
1
Front    ................................................................................................................................................................................................................
1
Cover    .......................................................................................................................................................................................................
2
Front Matter    .......................................................................................................

**Observations:**

1.   First 2 pages contains image or some watermarks text.
2.   Other 3 pages are index pages.



#### Checking the number of pages

In [8]:
len(manual)

4114

*  There are 4114 pages in the manual.

### Data Chunking

In [9]:
text_splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
    encoding_name='cl100k_base',
    chunk_size=512,
    chunk_overlap= 20
)

In [10]:
document_chunks = pdf_loader.load_and_split(text_splitter)

In [11]:
len(document_chunks)

8537

In [12]:
document_chunks[100].page_content

'cholecystokinin (which produces satiation), and increased leptin (an anorectic hormone produced by fat\ncells). Diminished taste and smell can decrease eating pleasure but usually decrease food intake only\nslightly. Anorexia may have other causes (eg, loneliness, inability to shop or prepare meals, dementia,\nsome chronic disorders, use of certain drugs). Depression is a common cause. Occasionally, anorexia\nnervosa (sometimes called anorexia tardive in the elderly), paranoia, or mania interferes with eating.\nDental problems limit the ability to chew and subsequently to digest foods. Swallowing difficulties (eg, due\nto strokes, other neurologic disorders, esophageal candidiasis, or xerostomia) are common. Poverty or\nfunctional impairment limits access to nutrients.\nThe institutionalized elderly are at particular risk of PEU. They are often confused and may be unable to\nexpress hunger or preferences for foods. They may be physically unable to feed themselves. Chewing or\nswallowi

In [ ]:
document_chunks[101].page_content

'compatible with good health). Vegans may develop vitamin B12 deficiency unless they consume yeast\nextracts or Asian-style fermented foods. Their intake of Ca, iron, and zinc also tends to be low. A fruit-only\ndiet is not recommended because it is deficient in protein, Na, and many micronutrients.\nFad diets: Some fad diets result in vitamin, mineral, and protein deficiencies; cardiac, renal, and\nmetabolic disorders; and sometimes death. Very low calorie diets (< 400 kcal/day) cannot sustain health\nfor long.\nDrugs and nutritional supplements: Many drugs (eg, appetite suppressants, digoxin) decrease\nappetite; others impair nutrient absorption or metabolism. Some drugs (eg, stimulants) have catabolic\neffects. Certain drugs can impair absorption of many nutrients; eg, anticonvulsants can impair absorption\nof vitamins.\nAlcohol or drug dependency: Patients with alcohol or drug dependency may neglect their nutritional\nneeds. Absorption and metabolism of nutrients may also be impair

**Observations:**

*   Generated the chunks using 512 chunk size and overlap size of 20.
*   There are 8537 chunks created with those paramters.
*   When checked the page and checked the content of chunk 100 and 101, It is observed that there is no overlapping of text that may be the blank space in the texts.



Lets increase the Check size to get overlapping of texts in chunks.

In [13]:
text_splitter_1 = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
    encoding_name='cl100k_base',
    chunk_size=512,
    chunk_overlap= 50
)

In [14]:
document_chunks_1 = pdf_loader.load_and_split(text_splitter_1)

In [15]:
len(document_chunks)

8537

In [ ]:
document_chunks_1[101].page_content

'cholecystokinin (which produces satiation), and increased leptin (an anorectic hormone produced by fat\ncells). Diminished taste and smell can decrease eating pleasure but usually decrease food intake only\nslightly. Anorexia may have other causes (eg, loneliness, inability to shop or prepare meals, dementia,\nsome chronic disorders, use of certain drugs). Depression is a common cause. Occasionally, anorexia\nnervosa (sometimes called anorexia tardive in the elderly), paranoia, or mania interferes with eating.\nDental problems limit the ability to chew and subsequently to digest foods. Swallowing difficulties (eg, due\nto strokes, other neurologic disorders, esophageal candidiasis, or xerostomia) are common. Poverty or\nfunctional impairment limits access to nutrients.\nThe institutionalized elderly are at particular risk of PEU. They are often confused and may be unable to\nexpress hunger or preferences for foods. They may be physically unable to feed themselves. Chewing or\nswallowi

In [ ]:
document_chunks_1[102].page_content

'and anorexia.\nVegetarian diets: Iron deficiency can occur in ovo-lacto vegetarians (although such a diet can be\ncompatible with good health). Vegans may develop vitamin B12 deficiency unless they consume yeast\nextracts or Asian-style fermented foods. Their intake of Ca, iron, and zinc also tends to be low. A fruit-only\ndiet is not recommended because it is deficient in protein, Na, and many micronutrients.\nFad diets: Some fad diets result in vitamin, mineral, and protein deficiencies; cardiac, renal, and\nmetabolic disorders; and sometimes death. Very low calorie diets (< 400 kcal/day) cannot sustain health\nfor long.\nDrugs and nutritional supplements: Many drugs (eg, appetite suppressants, digoxin) decrease\nappetite; others impair nutrient absorption or metabolism. Some drugs (eg, stimulants) have catabolic\neffects. Certain drugs can impair absorption of many nutrients; eg, anticonvulsants can impair absorption\nof vitamins.\nAlcohol or drug dependency: Patients with alcohol 

Observations:

*  After increasing the overlap size to 50, It is ovserved that there is overlap between the chunks.
*  There are few secnarios, when there is page break/end, the footer occupy the overlap size and the next chunk do not get the overlap text.
*  For this project lets go with overlap size of 50.

### Embedding

In [16]:
embedding_model = SentenceTransformerEmbeddings(model_name='thenlper/gte-large')

/tmp/ipykernel_29575/4198310515.py:1: LangChainDeprecationWarning: The class `HuggingFaceEmbeddings` was deprecated in LangChain 0.2.2 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-huggingface package and should be used instead. To use it run `pip install -U :class:`~langchain-huggingface` and import as `from :class:`~langchain_huggingface import HuggingFaceEmbeddings``.
  embedding_model = SentenceTransformerEmbeddings(model_name='thenlper/gte-large')
/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datas

modules.json:   0%|          | 0.00/385 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/57.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/619 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/670M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/342 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/191 [00:00<?, ?B/s]

In [ ]:
embedding_1 = embedding_model.embed_query(document_chunks_1[0].page_content)
embedding_2 = embedding_model.embed_query(document_chunks_1[1].page_content)

In [ ]:
print("Dimension of the embedding vector ",len(embedding_1))
len(embedding_1)==len(embedding_2)

Dimension of the embedding vector  1024


True

*  Verified that the embedding dimensions for chunk 0 and chunk 1 are both 1024, confirming that the embeddings were generated correctly.

### Vector Database

*  Choosing the Chroma for Vector Database.

In [17]:
out_dir = 'healthcare_assistance_db'

if not os.path.exists(out_dir):
  os.makedirs(out_dir)

In [18]:
vectorstore = Chroma.from_documents(
    document_chunks_1,
    embedding_model,
    persist_directory=out_dir
)

KeyboardInterrupt: 

In [ ]:
vectorstore = Chroma(persist_directory=out_dir,embedding_function=embedding_model)

/tmp/ipykernel_10206/2756559696.py:1: LangChainDeprecationWarning: The class `Chroma` was deprecated in LangChain 0.2.9 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-chroma package and should be used instead. To use it run `pip install -U :class:`~langchain-chroma` and import as `from :class:`~langchain_chroma import Chroma``.
  vectorstore = Chroma(persist_directory=out_dir,embedding_function=embedding_model)


### Retriever

In [ ]:
retriever = vectorstore.as_retriever(
    search_type='similarity',
    search_kwargs={'k': 2}
)

NameError: name 'vectorstore' is not defined

In [ ]:
rel_docs = retriever.get_relevant_documents("What is the protocol for managing sepsis in a critical care unit?")
rel_docs

/tmp/ipykernel_13262/1655728131.py:1: LangChainDeprecationWarning: The method `BaseRetriever.get_relevant_documents` was deprecated in langchain-core 0.1.46 and will be removed in 1.0. Use :meth:`~invoke` instead.
  rel_docs = retriever.get_relevant_documents("What is the protocol for managing sepsis in a critical care unit?")


[Document(metadata={'creator': 'Atop CHM to PDF Converter', 'creationDate': 'D:20120615054440Z', 'total_pages': 4114, 'moddate': '2026-03-17T23:16:37+00:00', 'author': '', 'trapped': '', 'source': '/content/drive/MyDrive/MedicalAssistant/medical_diagnosis_manual.pdf', 'format': 'PDF 1.7', 'keywords': '', 'modDate': 'D:20260317231637Z', 'subject': '', 'producer': 'pdf-lib (https://github.com/Hopding/pdf-lib)', 'title': 'The Merck Manual of Diagnosis & Therapy, 19th Edition', 'page': 2400, 'creationdate': '2012-06-15T05:44:40+00:00', 'file_path': '/content/drive/MyDrive/MedicalAssistant/medical_diagnosis_manual.pdf'}, page_content="16 - Critical Care Medicine\nChapter 222. Approach to the Critically Ill Patient\nIntroduction\nCritical care medicine specializes in caring for the most seriously ill patients. These patients are best\ntreated in an ICU staffed by experienced personnel. Some hospitals maintain separate units for special\npopulations (eg, cardiac, surgical, neurologic, pediatr

*  It is observed that for the given question, retriver found the 2 context besed on smiliarity search.

### System and User Prompt Template

In [ ]:
qna_system_message = """
You are an assistant whose work is to review the report and provide the appropriate answers from the context.
User input will have the context required by you to answer user questions.
This context will begin with the token: ###Context.
The context contains references to specific portions of a document relevant to the user query.

User questions will begin with the token: ###Question.

Please answer only using the context provided in the input. Do not mention anything about the context in your final answer.

If the answer is not found in the context, respond "I don't know".
"""

In [ ]:
qna_user_message_template = """
###Context
Here are some documents that are relevant to the question mentioned below.
{context}

###Question
{question}
"""

### Response Function

In [ ]:
def generate_rag_response(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=k)
    context_list = [d.page_content for d in relevant_document_chunks]

    # Combine document chunks into a single context
    context_for_query = ". ".join(context_list)

    user_message = qna_user_message_template.replace('{context}', context_for_query)
    user_message = user_message.replace('{question}', user_input)

    prompt = qna_system_message + '\n' + user_message

    # Generate the response
    try:
        response = llm(
                  prompt=prompt,
                  max_tokens=max_tokens,
                  temperature=temperature,
                  top_p=top_p,
                  top_k=top_k
                  )

        # Extract and print the model's response
        response = response['choices'][0]['text'].strip()
    except Exception as e:
        response = f'Sorry, I encountered the following error: \n {e}'

    return response

## Question Answering using RAG

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
user_input = "What is the protocol for managing sepsis in a critical care unit?"
print(generate_llama_response(user_input))

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
user_input = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
print(generate_rag_response(user_input))

Llama.generate: prefix-match hit


Based on the context provided, the answer would be: Appendicitis is characterized by abdominal pain, anorexia, and abdominal tenderness. It cannot be cured via medicine alone; instead, surgical removal of the appendix (appendectomy) is required for treatment.


### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
user_input = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
print(generate_rag_response(user_input))

Llama.generate: prefix-match hit


Based on the context provided, the condition being described is alopecia areata. The treatment options for this condition include topical, intralesional, or systemic corticosteroids, topical minoxidil, topical anthralin, topical immunotherapy (diphencyprone or squaric acid dibutylester), or psoralen plus ultraviolet A (PUVA). The cause of alopecia areata is believed to be an autoimmune disorder affecting genetically susceptible people exposed to unclear environmental triggers. Therefore, the possible


### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
user_input = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
print(generate_rag_response(user_input),max_tokens = 350)

Llama.generate: prefix-match hit


Based on the context provided, the recommended treatments for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function include:
1. Ensuring a reliable airway and maintaining adequate ventilation, oxygenation, and blood pressure.
2. Surgery to place monitors to track and treat intracranial pressure, decompress the brain if intracranial pressure is increased, or remove intracranial hematomas.
3. Supportive care, which includes preventing systemic complications due to immobilization, providing good nutrition, and preventing pressure


### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
user_input = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
print(generate_rag_response(user_input))

Llama.generate: prefix-match hit


Based on the context provided, here's the answer:

The person with a fractured leg should follow these steps:
1. Elevate the injured limb above heart level for the first 2 days to minimize swelling.
2. Apply warmth (using a heating pad) for 15 to 20 minutes after the first 48 hours to relieve pain and speed healing.
3. Immobilize the joints proximal and distal to the injury using elastic bandages, a cast, or a splint.
4. Keep the cast dry and never put


**Observations:**

*   List item
*   List item



### Fine-tuning

In [ ]:
user_input = "What is the protocol for managing sepsis in a critical care unit?"
print(generate_rag_response(user_input,max_tokens= 100))

Llama.generate: prefix-match hit


Based on the context, the protocol for managing sepsis in a critical care unit includes:
1. Prompt empiric antibiotic therapy, started immediately after suspecting sepsis.
2. Selection of antibiotics based on suspected source, clinical setting, knowledge or suspicion of causative organisms and sensitivity patterns common to that specific inpatient unit, and previous culture results.
3. Continuation of antibiotics for at least 5 days after shock res


In [ ]:
user_input = "What is the protocol for managing sepsis in a critical care unit?"
print(generate_rag_response(user_input,max_tokens= 350,k=5))

In [ ]:
user_input = "What is the protocol for managing sepsis in a critical care unit?"
print(generate_rag_response(user_input,temparature=0.8,max_tokens=1024))

In [ ]:
user_input = "What is the protocol for managing sepsis in a critical care unit?"
print(generate_rag_response(user_input,temparature=0.9, k=10))

In [ ]:
user_input = "What is the protocol for managing sepsis in a critical care unit?"
print(generate_rag_response(user_input,top_p=0.1, top_k=20))

In [ ]:
user_input = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
print(generate_rag_response(user_input,temperature=0.1, max_tokens=350))

Llama.generate: prefix-match hit


Based on the context provided, the answer would be: Appendicitis is characterized by abdominal pain, anorexia, and abdominal tenderness. It cannot be cured via medicine alone; instead, surgical removal of the appendix (appendectomy) is required for treatment.


In [ ]:
user_input = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

Llama.generate: prefix-match hit


The common symptoms for appendicitis include epigastric or periumbilical pain followed by brief nausea, vomiting, and anorexia. After a few hours, the pain shifts to the right lower quadrant. Pain increases with cough and motion. Classic signs are right lower quadrant direct and rebound tenderness located at McBurney's point. Additional signs include pain felt in the right lower quadrant with palpation of the left lower quadrant (Rovsing sign), an increase in pain from passive extension of the right hip joint that stretches the iliopsoas muscle (psoas sign), or pain caused by passive internal rotation of the flexed thigh (obturator sign). Low-grade fever is common.

Appendicitis cannot be cured via medicine alone, and surgical removal is required for treatment. The surgical procedure for treating appendicitis is an open or laparoscopic appendectomy. If the appendix is perforated, antibiotics should be continued until the patient's temperature and WBC count have normalized or according 

In [ ]:
user_input = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

In [ ]:
user_input = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
print(generate_rag_response(user_input,k=5, temperature=0.1, max_tokens=350))

**Observations:**

*   List item
*   List item



## Output Evaluation

Let us now use the LLM-as-a-judge method to check the quality of the RAG system on two parameters - retrieval and generation. We illustrate this evaluation based on the answeres generated to the question from the previous section.

- We are using the same Mistral model for evaluation, so basically here the llm is rating itself on how well he has performed in the task.

In [ ]:
groundedness_rater_system_message = """
You are tasked with rating AI generated answers to questions posed by users.
You will be presented a question, context used by the AI system to generate the answer and an AI generated answer to the question.
In the input, the question will begin with ###Question, the context will begin with ###Context while the AI generated answer will begin with ###Answer.

Evaluation criteria:
The task is to judge the extent to which the metric is followed by the answer.
1 - The metric is not followed at all
2 - The metric is followed only to a limited extent
3 - The metric is followed to a good extent
4 - The metric is followed mostly
5 - The metric is followed completely

Metric:
The answer should be derived only from the information presented in the context

Instructions:
1. First write down the steps that are needed to evaluate the answer as per the metric.
2. Give a step-by-step explanation if the answer adheres to the metric considering the question and context as the input.
3. Next, evaluate the extent to which the metric is followed.
4. Use the previous information to rate the answer using the evaluaton criteria and assign a score.
"""

In [ ]:
relevance_rater_system_message = """
You are tasked with rating AI generated answers to questions posed by users.
You will be presented a question, context used by the AI system to generate the answer and an AI generated answer to the question.
In the input, the question will begin with ###Question, the context will begin with ###Context while the AI generated answer will begin with ###Answer.

Evaluation criteria:
The task is to judge the extent to which the metric is followed by the answer.
1 - The metric is not followed at all
2 - The metric is followed only to a limited extent
3 - The metric is followed to a good extent
4 - The metric is followed mostly
5 - The metric is followed completely

Metric:
Relevance measures how well the answer addresses the main aspects of the question, based on the context.
Consider whether all and only the important aspects are contained in the answer when evaluating relevance.

Instructions:
1. First write down the steps that are needed to evaluate the context as per the metric.
2. Give a step-by-step explanation if the context adheres to the metric considering the question as the input.
3. Next, evaluate the extent to which the metric is followed.
4. Use the previous information to rate the context using the evaluaton criteria and assign a score.
"""

In [ ]:
user_message_template = """
###Question
{question}

###Context
{context}

###Answer
{answer}
"""

In [ ]:
def generate_ground_relevance_response(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=3)
    context_list = [d.page_content for d in relevant_document_chunks]
    context_for_query = ". ".join(context_list)

    # Combine user_prompt and system_message to create the prompt
    prompt = f"""[INST]{qna_system_message}\n
                {'user'}: {qna_user_message_template.format(context=context_for_query, question=user_input)}
                [/INST]"""

    response = llm(
            prompt=prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    answer =  response["choices"][0]["text"]

    # Combine user_prompt and system_message to create the prompt
    groundedness_prompt = f"""[INST]{groundedness_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    # Combine user_prompt and system_message to create the prompt
    relevance_prompt = f"""[INST]{relevance_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    response_1 = llm(
            prompt=groundedness_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    response_2 = llm(
            prompt=relevance_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    return response_1['choices'][0]['text'],response_2['choices'][0]['text']

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
user_input = "What is the protocol for managing sepsis in a critical care unit?"
ground,rel = generate_ground_relevance_response(user_input,max_tokens=350)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Steps to evaluate the answer:
1. Identify the information in the context related to managing sepsis in a critical care unit.
2. Compare the information from the context with the AI generated answer to determine if the answer is derived only from the context.
3. Evaluate the extent to which the metric (answer should be derived only from the information presented in the context) is followed.

Explanation:
The AI generated answer includes several aspects of managing sepsis in a critical care unit, such as monitoring vital signs, quantifying fluid intake and output, measuring daily weight, providing adequate nutrition, routine blood tests, empiric antibiotic therapy, and normalization of blood glucose levels. These elements are all present in the context.

The answer is derived directly from the context, as it includes all the necessary information for managing sepsis in a critical care unit that is presented in the context. Therefore, the metric is followed completely.

Rating:
Based on 

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
user_input_2 = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
ground,rel = generate_ground_relevance_response(user_input_2,max_tokens=1024)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Steps to evaluate the answer:
1. Identify the information in the context related to appendicitis and its symptoms.
2. Determine if the AI generated answer includes only the information derived from the context.
3. Evaluate the extent to which the metric is followed.

The AI generated answer "The common symptoms for appendicitis include abdominal pain, anorexia, and abdominal tenderness. Appendicitis cannot be cured via medicine alone; the standard treatment is surgical removal of the appendix through open or laparoscopic appendectomy." adheres to the metric as it only includes information derived from the context about the symptoms of appendicitis and the fact that it cannot be cured with medicine, but requires surgical removal.

Therefore, the extent to which the metric is followed is 5 - The metric is followed completely.

Based on the evaluation criteria and the extent to which the metric is followed, I would rate the answer as a 5.

 Steps to evaluate the context as per the releva

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
user_input_3 = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
ground,rel = generate_ground_relevance_response(user_input_3,max_tokens=500)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Steps to evaluate the answer:
1. Identify the information related to effective treatments or solutions for sudden patchy hair loss from the context.
2. Compare the identified information with the AI generated answer to check if all the elements in the answer are derived only from the context.

Explanation:
The AI generated answer mentions the effective treatments or solutions for addressing sudden patchy hair loss as topical, intralesional, or systemic corticosteroids, topical minoxidil, topical anthralin, topical immunotherapy (diphencyprone or squaric acid dibutylester), or psoralen plus ultraviolet A (PUVA). These treatments are explicitly mentioned in the context. The cause of sudden patchy hair loss being alopecia areata is also correctly identified from the context and included in the answer. Therefore, the answer adheres to the metric as it is derived only from the information presented in the context.

Evaluation:
The metric is followed completely.

Rating:
Based on the evalua

### Query 4: What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
user_input_4 = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
ground,rel = generate_ground_relevance_response(user_input_3,max_tokens=500)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Steps to evaluate the answer:
1. Identify the information related to effective treatments or solutions for sudden patchy hair loss from the context.
2. Compare the identified information with the AI generated answer to check if all the elements in the answer are derived only from the context.

Explanation:
The AI generated answer mentions the effective treatments or solutions for addressing sudden patchy hair loss as topical, intralesional, or systemic corticosteroids, topical minoxidil, topical anthralin, topical immunotherapy (diphencyprone or squaric acid dibutylester), or psoralen plus ultraviolet A (PUVA). These treatments are explicitly mentioned in the context. The cause of sudden patchy hair loss being alopecia areata is also correctly identified from the context and included in the answer. Therefore, the answer adheres to the metric as it is derived only from the information presented in the context.

Evaluation:
The metric is followed completely.

Rating:
Based on the evalua

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
user_input_5 = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
ground,rel = generate_ground_relevance_response(user_input_3,max_tokens=500)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Steps to evaluate the answer:
1. Identify the information related to effective treatments or solutions for sudden patchy hair loss from the context.
2. Compare the identified information with the AI generated answer to check if all the elements in the answer are derived only from the context.

Explanation:
The AI generated answer mentions the effective treatments or solutions for addressing sudden patchy hair loss as topical, intralesional, or systemic corticosteroids, topical minoxidil, topical anthralin, topical immunotherapy (diphencyprone or squaric acid dibutylester), or psoralen plus ultraviolet A (PUVA). These treatments are explicitly mentioned in the context. The cause of sudden patchy hair loss being alopecia areata is also correctly identified from the context and included in the answer. Therefore, the answer adheres to the metric as it is derived only from the information presented in the context.

Evaluation:
The metric is followed completely.

Rating:
Based on the evalua

**Observations:**

*  For groundedness, the score is clearly stated (5) with a reason,the response appears fully derived from the context without missing information.

*  For relevance, the score is 4 as the answer follows the maric mostly but more information need to improve the answer.



## Actionable Insights and Business Recommendations



*   As we see that while chunking, need to increase the overlap size to 50 to ensure that the overlapping of text should be available in the chunks. This was missing due to footer and the blank spaces.
*   Changing the temparature values shows how the response gets change from most concise to generic/imaginative.
*   Max_token will play a crucial role, when higher values yield detailed responses, while simple queries result in concise outputs despite large token limits due to prompt design.
*  The Evaluation of the RAG show that it is stick to the Relevence and groundedness while providing the response.
*  Continuously adjust RAG parameters based on specific use cases for optimal performance.
*  Retrieval parameter k is critical as the answer can be spread across multiple contexts.



<font size=6 color='blue'>Power Ahead</font>
___